Firestore/Swift/Source/Codable/Timestamp+Codable.swift

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import FirebaseFirestore
+
+/**
+ * A protocol describing the encodable properties of a DocumentSnapshot.
+ *
+ * Note: this protocol exists as a workaround for the Swift compiler: if the Timestamp class
+ * was extended directly to conform to Codable, the methods implementing the protocol would be need
+ * to be marked required but that can't be done in an extension. Declaring the extension on the
+ * protocol sidesteps this issue.
+ */
+public protocol CodableTimestamp: Codable {
+  init(date: Date)
+  func dateValue() -> Date
```

Comment:
The properties on `CodableTimestamp` should reflect the data it contains and how it should be serialized with a non-Firestore encoder.

The whole reason Timestamp exists is because NSDate can't losslessly represent Firestore time data, so the properties here should be `seconds` and `nanoseconds`.

Note that the properties here are for the swift compiler to generate the Codable implementation. End users using Timestamps will already have access to these methods through `FIRTimestamp` so this init method and helper don't belong here.